joint_limits/src/joint_saturation_limiter.cpp

```diff
@@ -0,0 +1,451 @@
+      {
+        desired_acc[index] = std::copysign(
+          std::min(std::fabs(desired_acc[index]), joint_limits_[index].max_acceleration),
+          desired_acc[index]);
+      }
+
+      // Recompute velocity and position
+      if (has_vel_cmd)
+      {
+        desired_vel[index] =
+          current_joint_states.velocities[index] + desired_acc[index] * dt_seconds;
+      }
+      if (has_pos_cmd)
+        desired_pos[index] = current_joint_states.positions[index] +
+                             current_joint_states.velocities[index] * dt_seconds +
+                             0.5 * desired_acc[index] * dt_seconds * dt_seconds;
+    }
+    std::ostringstream ostr;
+    for (auto jnt : limited_jnts_pos) ostr << jnt << " ";
+    ostr << "\b \b";  // erase last character
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_logging_itf_->get_logger(), *clock_, ROS_LOG_THROTTLE_PERIOD,
+      "Joint(s) [" << ostr.str().c_str()
+                   << "] would exceed position limits"
+                      " if continuing at current state, limiting all joints");
+  }
+
+  // display limitations
+
+  // if position limiting
+  if (limited_jnts_pos.size() > 0)
+  {
+    std::ostringstream ostr;
+    for (auto jnt : limited_jnts_pos) ostr << jnt << " ";
+    ostr << "\b \b";  // erase last character
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_logging_itf_->get_logger(), *clock_, ROS_LOG_THROTTLE_PERIOD,
+      "Joint(s) [" << ostr.str().c_str() << "] would exceed position limits, limiting");
+  }
+
+  if (limited_jnts_vel.size() > 0)
+  {
+    std::ostringstream ostr;
+    for (auto jnt : limited_jnts_vel) ostr << jnt << " ";
+    ostr << "\b \b";  // erase last character
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_logging_itf_->get_logger(), *clock_, ROS_LOG_THROTTLE_PERIOD,
+      "Joint(s) [" << ostr.str().c_str() << "] would exceed velocity limits, limiting");
+  }
+
+  if (limited_jnts_acc.size() > 0)
+  {
+    std::ostringstream ostr;
+    for (auto jnt : limited_jnts_acc) ostr << jnt << " ";
+    ostr << "\b \b";  // erase last character
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_logging_itf_->get_logger(), *clock_, ROS_LOG_THROTTLE_PERIOD,
+      "Joint(s) [" << ostr.str().c_str() << "] would exceed acceleration limits, limiting");
+  }
+
+  if (limited_jnts_dec.size() > 0)
+  {
+    std::ostringstream ostr;
+    for (auto jnt : limited_jnts_dec) ostr << jnt << " ";
+    ostr << "\b \b";  // erase last character
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_logging_itf_->get_logger(), *clock_, ROS_LOG_THROTTLE_PERIOD,
+      "Joint(s) [" << ostr.str().c_str() << "] would exceed deceleration limits, limiting");
+  }
+
+  if (has_pos_cmd) desired_joint_states.positions = desired_pos;
+  if (has_vel_cmd) desired_joint_states.velocities = desired_vel;
+  if (has_acc_cmd) desired_joint_states.accelerations = desired_acc;
+
+  return limits_enforced;
+}
+
+template <>
+bool JointSaturationLimiter<JointLimits>::on_enforce_effort(
```

Comment:
I purpose to keep this method even more generic, but always using effort limits. Effort can mean a lot therefore let's add vectors as input/output and not JointTrajectoryPoint. Then keep the same name as the above method.